pkg/controller/build/controller_test.go

```diff
@@ -95,12 +100,203 @@ func (f *fixture) updateIndex(i informers.SharedInformerFactory, bl []*v1alpha1.
 func getKey(build *v1alpha1.Build, t *testing.T) string {
 	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(build)
 	if err != nil {
-		t.Errorf("Unexpected error getting key for build %v: %v", build.Name, err)
+		t.Errorf("unexpected error getting key for build %v: %v", build.Name, err)
 		return ""
 	}
 	return key
 }
 
+func TestBuildNotFoundFlow(t *testing.T) {
+	bldr := &nop.Builder{}
+
+	build := newBuild("test")
+	f := &fixture{
+		t:          t,
+		objects:    []runtime.Object{build},
+		client:     fake.NewSimpleClientset(build),
+		kubeclient: k8sfake.NewSimpleClientset(),
+	}
+
+	stopCh := make(chan struct{})
+	eventCh := make(chan string, 1024)
+	defer close(stopCh)
+	defer close(eventCh)
+
+	c, i, k8sI := f.newController(bldr, eventCh)
+	f.updateIndex(i, []*v1alpha1.Build{build})
+	i.Start(stopCh)
+	k8sI.Start(stopCh)
+
+	// induce failure when fetching build information in controller
+	reactor := func(action clientgotesting.Action) (handled bool, ret runtime.Object, err error) {
+		if action.GetVerb() == "get" && action.GetResource().Resource == "builds" {
+			return true, nil, fmt.Errorf("Inducing failure for %q action of %q", action.GetVerb(), action.GetResource().Resource)
+		}
+		return false, nil, nil
+	}
+	f.client.PrependReactor("*", "*", reactor)
+
+	if err := c.syncHandler(getKey(build, t)); err == nil {
+		t.Errorf("Expect error syncing build")
+	}
+}
+
+func TestBuildWithBadKey(t *testing.T) {
+	bldr := &nop.Builder{}
+
+	f := &fixture{
+		t:          t,
+		kubeclient: k8sfake.NewSimpleClientset(),
+	}
+	eventCh := make(chan string, 1024)
+	c, _, _ := f.newController(bldr, eventCh)
+
+	if err := c.syncHandler("bad/worse/worst"); err != nil {
+		t.Errorf("Unexpected error while syncing build: %s", err.Error())
+	}
+}
+
+func TestBuildNotFoundError(t *testing.T) {
+	bldr := &nop.Builder{}
+
+	build := newBuild("test")
+	f := &fixture{
+		t:          t,
+		objects:    []runtime.Object{build},
+		client:     fake.NewSimpleClientset(build),
+		kubeclient: k8sfake.NewSimpleClientset(),
+	}
+
+	stopCh := make(chan struct{})
+	eventCh := make(chan string, 1024)
+	defer close(stopCh)
+	defer close(eventCh)
+
+	c, i, k8sI := f.newController(bldr, eventCh)
+	// Don't update build informers with test build object
+	i.Start(stopCh)
+	k8sI.Start(stopCh)
+
+	if err := c.syncHandler(getKey(build, t)); err != nil {
+		t.Errorf("Unexpected error while syncing build: %s", err.Error())
+	}
+}
+
+func TestBuildWithNonExistentTemplates(t *testing.T) {
+	tests := []struct {
+		bldr  builder.Interface
```

Comment:
`bldr` is only passed as `&nop.Builder{}`, so it doesn't need to be a field in the test case struct.

In fact, the only thing that varies between the two `build`s is the `Kind`, so maybe just make that the test case:

```
for _, kind := []string{BuildTemplateKind, ClusterBuildTemplateKind} {
  ...
}
```